### PR TITLE
Update Nautobot v2.0.0rc2

### PR DESCRIPTION
**Pull request by snaselj**
# Closes NaN

## What's Changed

- Updated `nautobot` to `v2.0.0rc2`.

## To Do

- [ ] Remove `TBD`s from code.
- [ ] Add better CI version matrix for testing, currently only `3.8` and `2.0.0rc2` is tested.
- [ ] Improve documentation.

### CODE REVIEW

**Comment by snaselj** [comment]
Ready for review now, just a few points to resolve described in this PR notes.

File: docs/admin/uninstall.md
Comment (glennmatthews) [inline docs/admin/uninstall.md:1]: Not sure about the changes to this file - I'm pretty sure the existing recommendations about removing ObjectChanges and Status associations are still correct and relevant.

File: docs/dev/dev_environment.md [hunk 1]
```diff
@@ -101,9 +101,6 @@ The project features a CLI helper based on [Invoke](https://www.pyinvoke.org/) t
 
 Each command can be executed with `invoke <command>`. All commands support the arguments `--nautobot-ver` and `--python-ver` if you want to manually define the version of Python and Nautobot to use. Each command also has its own help `invoke <command> --help`
 
-!!! note
-    To run the mysql (mariadb) development environment, set the environment variable as such `export NAUTOBOT_USE_MYSQL=1`.
```
Comment (glennmatthews) [inline docs/dev/dev_environment.md:105]: No longer accurate/relevant?

File: nautobot_floor_plan/__init__.py [hunk 2]
```diff
@@ -1,18 +1,14 @@
...
-    from importlib import metadata
-except ImportError:
-    # Python version < 3.8
-    import importlib_metadata as metadata
+from importlib import metadata
 
 __version__ = metadata.version(__name__)
 
-from nautobot.extras.plugins import PluginConfig
+from nautobot.extras.plugins import NautobotAppConfig
```
Comment (glennmatthews) [inline nautobot_floor_plan/__init__.py:7]:
```suggestion
from nautobot.apps import NautobotAppConfig
```

File: nautobot_floor_plan/api/serializers.py [hunk 3]
```diff
@@ -1,34 +1,22 @@
...
-
+from nautobot.core.api.serializers import NautobotModelSerializer
+from nautobot.extras.api.mixins import TaggedModelSerializerMixin
 from nautobot_floor_plan import models
 
-from .nested_serializers import *  # noqa: F403, pylint: disable=unused-import,unused-wildcard-import,wildcard-import
-
 
-class FloorPlanSerializer(NautobotModelSerializer, TaggedObjectSerializer):
+class FloorPlanSerializer(NautobotModelSerializer):
```
Comment (glennmatthews) [inline nautobot_floor_plan/api/serializers.py:7]: Probably also needs `TaggedModelSerializerMixin`.

File: nautobot_floor_plan/api/urls.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 """Django API urlpatterns declaration for nautobot_floor_plan plugin."""
 
-from nautobot.core.api import OrderedDefaultRouter
+from nautobot.core.api.routers import OrderedDefaultRouter
```
Comment (glennmatthews) [inline nautobot_floor_plan/api/urls.py:3]:
```suggestion
from nautobot.apps.api import OrderedDefaultRouter
```

File: nautobot_floor_plan/choices.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 """ChoiceSet definitions for nautobot_floor_plan."""
 
-from nautobot.utilities.choices import ChoiceSet
+from nautobot.core.choices import ChoiceSet
```
Comment (glennmatthews) [inline nautobot_floor_plan/choices.py:3]:
```suggestion
from nautobot.apps.choices import ChoiceSet
```

File: nautobot_floor_plan/filters.py [hunk 6]
```diff
@@ -4,7 +4,7 @@
 
 from nautobot.dcim.models import Location, Rack
 from nautobot.extras.filters import NautobotFilterSet
-from nautobot.utilities.filters import NaturalKeyOrPKMultipleChoiceFilter, SearchFilter, TagFilter
+from nautobot.core.filters import NaturalKeyOrPKMultipleChoiceFilter, SearchFilter, TagFilter
```
Comment (glennmatthews) [inline nautobot_floor_plan/filters.py:7]:
```suggestion
from nautobot.apps.filters import NaturalKeyOrPKMultipleChoiceFilter, SearchFilter
from nautobot.core.filters import TagFilter
```

Though `TagFilter` may not be necessary per se as the filterset **should** automatically have a `tags` (not `tag`) filter field?

File: nautobot_floor_plan/forms.py [hunk 7]
```diff
@@ -1,16 +1,18 @@
...
 from nautobot.extras.forms import (
-    CustomFieldModelCSVForm,
     NautobotBulkEditForm,
     NautobotFilterForm,
     NautobotModelForm,
     TagsBulkEditFormMixin,
 )
-from nautobot.utilities.forms import (
-    CSVModelChoiceField,
+from nautobot.core.forms import (
```
Comment (glennmatthews) [inline nautobot_floor_plan/forms.py:15]:
```suggestion
from nautobot.apps.forms import (
```

File: nautobot_floor_plan/models.py [hunk 8]
```diff
@@ -194,23 +171,10 @@ def clean(self):
             # Else they must overlap
             raise ValidationError("Tile overlaps with another defined tile.")
 
-    def get_absolute_url(self):
+    def get_absolute_url(self, api=False):
         """Return detail view for FloorPlanTile."""
         return reverse("plugins:nautobot_floor_plan:floorplantile", args=[self.pk])
```
Comment (glennmatthews) [inline nautobot_floor_plan/models.py:176]: Remove?

File: nautobot_floor_plan/tests/test_api_views.py [hunk 9]
```diff
@@ -1,10 +1,11 @@
 """Unit tests for nautobot_floor_plan."""
+import unittest
```
Comment (glennmatthews) [inline nautobot_floor_plan/tests/test_api_views.py:2]: Why's this needed?

File: pyproject.toml [hunk 10]
```diff
@@ -17,39 +17,33 @@ packages = [
 ]
 
 [tool.poetry.dependencies]
-python = "^3.7"
-# Required for Python 3.7 for now. See: https://stackoverflow.com/a/73932581/194311
-importlib-metadata = "4.13.0"
-# Used for local development
-nautobot = { version = "^1.4.0", optional = true }
+nautobot = { version = ">=2.0.0-rc.1", optional = true }
```
Comment (glennmatthews) [inline pyproject.toml:20]:
```suggestion
nautobot = { version = ">=2.0.0-rc.2", optional = true }
```

**Comment by glennmatthews** [comment]
```
❯ poetry install             
Installing dependencies from lock file
Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.

Because nautobot-floor-plan depends on nautobot (>=2.0.0-rc.1) which doesn't match any versions, version solving failed.
```

**Comment by glennmatthews** [comment]
`poetry add 'nautobot>=2.0.0rc2'`appears to fix the above for me.

File: nautobot_floor_plan/template_content.py [hunk 11]
```diff
@@ -21,7 +21,7 @@ def buttons(self):
                 return ""
 
             delete_url = reverse("plugins:nautobot_floor_plan:floorplan_delete", kwargs={"pk": floor_plan.pk})
-            return_url = reverse("dcim:location", kwargs={"slug": location.slug})
+            return_url = reverse("dcim:location", kwargs={"name": location.name})
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot_floor_plan/template_content.py:24]: As currently coded this raises a NoReverseMatch exception as the location URL expects a pk, not a name, in 2.0.

File: docs/dev/dev_environment.md (hunk 1 above)
Comment (snaselj) [reply docs/dev/dev_environment.md:105]: This comes from the template, `NAUTOBOT_USE_MYSQL` is not in the repo anywhere. To use mysql you can `cp invoke.mysql.yml invoke.yml`.

File: nautobot_floor_plan/models.py [hunk 12]
```diff
@@ -8,8 +8,8 @@
 from django.urls import reverse
 
 from nautobot.core.models.generics import PrimaryModel
-from nautobot.extras.models import StatusModel
 from nautobot.extras.utils import extras_features
+from nautobot.extras.models import StatusField
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot_floor_plan/models.py]: Fixed

File: nautobot_floor_plan/models.py [hunk 13]
```diff
@@ -102,9 +83,12 @@ def get_svg(self, *, user, base_url):
     "statuses",
     "webhooks",
 )
-class FloorPlanTile(PrimaryModel, StatusModel):
+# TBD: Remove after releasing pylint-nautobot v0.3.0
+# pylint: disable-next=nb-string-field-blank-null
+class FloorPlanTile(PrimaryModel):
     """Model representing a single rectangular "tile" within a FloorPlan, its status, and any Rack that it contains."""
 
+    status = StatusField(null=True)
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot_floor_plan/models.py:91]: Without `null=True`, migrations were failing. What should be the default status here? Probably need to refactor migrations to create (or update) the default status value in the migration prior to one altering this field.

File: nautobot_floor_plan/tests/test_api_views.py [hunk 14]
```diff
@@ -1,10 +1,11 @@
...
 
 from django.contrib.contenttypes.models import ContentType
 
 from nautobot.dcim.models import Rack
 from nautobot.extras.models import Tag
-from nautobot.utilities.testing import APIViewTestCases
+from nautobot.core.testing import APIViewTestCases
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot_floor_plan/tests/test_api_views.py]: Fixed

File: nautobot_floor_plan/tests/test_api_views.py (hunk 9 above)
Comment (snaselj) [reply nautobot_floor_plan/tests/test_api_views.py]: Removed

File: pyproject.toml (hunk 10 above)
Comment (snaselj) [reply pyproject.toml:20]: Works for me with the current value `rc.1`. Could you pls. check there is no `invoke.yml` on your localhost or something similar?

**Comment by snaselj** [comment]
> ```
> ❯ poetry install             
> Installing dependencies from lock file
> Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.
> 
> Because nautobot-floor-plan depends on nautobot (>=2.0.0-rc.1) which doesn't match any versions, version solving failed.
> ```

Sorry for this, updated the lock file.

File: docs/admin/uninstall.md
Comment (snaselj) [reply docs/admin/uninstall.md:1]: https://github.com/nautobot/cookiecutter-nautobot-app/blob/develop/nautobot-app/%7B%7B%20cookiecutter.project_slug%20%7D%7D/docs/admin/uninstall.md

This comes from the template.

@cmsirbu Should I add the following back to the template?

https://github.com/nautobot/nautobot-plugin-floor-plan/pull/46/files#diff-f3d884bd1311b1c18fe9964a97e211caa0c8f2d9cd43335e30055ca21f421b92L7-L31

Comment (bryanculver) [reply docs/admin/uninstall.md:1]: Let's crib from: https://github.com/nautobot/nautobot/blob/01d077e393401fa9faa6500d96f16ef2eeeafd16/nautobot/dcim/migrations/0046_fixup_null_statuses.py

File: nautobot_floor_plan/models.py [hunk 15]
```diff
@@ -61,32 +60,14 @@ class Meta:
...
-        "x_size",
-        "y_size",
-        "tile_width",
-        "tile_depth",
-    ]
-
-    def get_absolute_url(self):
+    def get_absolute_url(self, api=False):
         """Return detail view for FloorPlan."""
         return reverse("plugins:nautobot_floor_plan:floorplan", args=[self.pk])
```
[the remark this answers is not in the document]
Comment (bryanculver) [reply nautobot_floor_plan/models.py]: @snaselj Please remove

File: nautobot_floor_plan/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (bryanculver) [reply nautobot_floor_plan/models.py:86]: We're going to crib from Core on how to fix up the null statuses and make it `blank=False, null=False`

File: nautobot_floor_plan/models.py (hunk 8 above)
Comment (bryanculver) [reply nautobot_floor_plan/models.py]: @snaselj please remove

File: nautobot_floor_plan/migrations/0002_auto_20230901_0845.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
+# Generated by Django 3.2.20 on 2023-09-01 08:45
+
+from django.db import migrations, models
+import django.db.models.deletion
+import nautobot.core.models.fields
+import nautobot.extras.models.statuses
+
+
+class Migration(migrations.Migration):
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot_floor_plan/migrations/0002_auto_20230901_0845.py]: Fixed

File: nautobot_floor_plan/filters.py [hunk 17]
```diff
@@ -15,13 +16,13 @@ class FloorPlanFilterSet(NautobotFilterSet):
...
-            "location__slug": "icontains",
         },
     )
     location = NaturalKeyOrPKMultipleChoiceFilter(
         queryset=Location.objects.all(),
-        label="Location (slug or ID)",
+        label="Location (name or ID)",
     )
+    # TBD: Verify it's necessary
     tag = TagFilter()
```
Comment (bryanculver) [inline nautobot_floor_plan/filters.py:26]:
```suggestion
```

Add `"tags"` to fields in Meta below. Same for `FloorPlanTileFilterSet`.

File: nautobot_floor_plan/forms.py [hunk 18]
```diff
@@ -72,7 +62,7 @@ class FloorPlanFilterForm(NautobotFilterForm):
     field_order = ["q", "location", "x_size", "y_size"]
 
     q = forms.CharField(required=False, label="Search")
-    location = DynamicModelMultipleChoiceField(queryset=Location.objects.all(), to_field_name="slug", required=False)
+    location = DynamicModelMultipleChoiceField(queryset=Location.objects.all(), to_field_name="name", required=False)
```
Comment (bryanculver) [inline nautobot_floor_plan/forms.py:65]: `name` is not unique. I think we need to use `pk` here.
```suggestion
    location = DynamicModelMultipleChoiceField(queryset=Location.objects.all(), to_field_name="pk", required=False)
```

File: pyproject.toml [hunk 19]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "nautobot-floor-plan"
-version = "1.0"
+version = "2.0.0rc2"
```
Comment (bryanculver) [inline pyproject.toml:3]:
```suggestion
version = "2.0.0rc1"
```

We haven't shipped an RC1 of this app yet.

File: pyproject.toml [hunk 20]
```diff
@@ -17,39 +17,33 @@ packages = [
 ]
 
 [tool.poetry.dependencies]
-python = "^3.7"
-# Required for Python 3.7 for now. See: https://stackoverflow.com/a/73932581/194311
-importlib-metadata = "4.13.0"
-# Used for local development
-nautobot = { version = "^1.4.0", optional = true }
+nautobot = { version = "^2.0.0rc1", optional = true }
```
Comment (bryanculver) [inline pyproject.toml:20]:
```suggestion
nautobot = { version = ">=2.0.0rc1,<2.0.0rc99" }
```

File: pyproject.toml [hunk 21]
```diff
@@ -17,39 +17,33 @@ packages = [
...
 [tool.poetry.dev-dependencies]
+Markdown = "*"
 bandit = "*"
 black = "*"
 coverage = "*"
 django-debug-toolbar = "*"
 # we need to pin flake8 because of package dependencies that cause it to downgrade and
 # therefore cause issues with linting since older versions do not take .flake8 as config
 flake8 = "^3.9.2"
+griffe = "0.30.1"               # Last version with python 3.7 support
```
Comment (bryanculver) [inline pyproject.toml:32]: Dropping 3.7 support this and similar can be removed.

**Review by bryanculver — CHANGES_REQUESTED** [review]
Almost there @snaselj. Just a few cleanups and then we can ship an RC1

File: nautobot_floor_plan/filters.py (hunk 17 above)
Comment (snaselj) [reply nautobot_floor_plan/filters.py]: When I removed this line and added it to Meta.fields, tests were complaining tag is missing in the model, so keeping it as is, just removed TBDs.

File: nautobot_floor_plan/forms.py (hunk 18 above)
Comment (snaselj) [reply nautobot_floor_plan/forms.py]: Fixed

File: pyproject.toml (hunk 19 above)
Comment (snaselj) [reply pyproject.toml]: Fixed

File: pyproject.toml (hunk 20 above)
Comment (snaselj) [reply pyproject.toml]: Fixed

File: pyproject.toml (hunk 21 above)
Comment (snaselj) [reply pyproject.toml]: Removed comments.

**Comment by snaselj** [comment]
> Almost there @snaselj. Just a few cleanups and then we can ship an RC1

PR comments are addressed, ready to review and ship.

File: nautobot_floor_plan/filters.py (hunk 17 above)
Comment (snaselj) [reply nautobot_floor_plan/filters.py]: Fixed now by removing `tag` filter and replacing with `tags`.

**Review by bryanculver — APPROVED** [review]

**Pull request merged**